<a href="https://colab.research.google.com/github/osvaldofacundo/proyectowebscraping/blob/main/webscraping_fixture_arg2022.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
#llamamos las librerias 
from bs4 import BeautifulSoup
import requests
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import re
import plotly.express as px
import statsmodels.api as sm
import statsmodels.formula.api as smf
from scipy.stats import poisson

In [2]:
webpage= requests.get('https://ar.marca.com/claro/futbol/primera-division/fixture.html') #la página de donde sacaremos la data
soup = BeautifulSoup(webpage.content, "html.parser") #descargamos el html

In [3]:
# solo 18 fechas que son las que se jugaron y estan completas a la fecha 16/09/22
fecha1= soup.find_all(id="jornada1") # identificamos la tabla y buscams y extraemos ella del codigo html
fecha2= soup.find_all(id="jornada2")
fecha3= soup.find_all(id="jornada3")
fecha4= soup.find_all(id="jornada4")
fecha5= soup.find_all(id="jornada5")
fecha6= soup.find_all(id="jornada6")
fecha7= soup.find_all(id="jornada7")
fecha8= soup.find_all(id="jornada8")
fecha9= soup.find_all(id="jornada9")
fecha10= soup.find_all(id="jornada10")
fecha11= soup.find_all(id="jornada11")
fecha12= soup.find_all(id="jornada12")
fecha13= soup.find_all(id="jornada13")
fecha14= soup.find_all(id="jornada14")
fecha15= soup.find_all(id="jornada15")
fecha16= soup.find_all(id="jornada16")
fecha17= soup.find_all(id="jornada17")
fecha18= soup.find_all(id="jornada18")
fecha19= soup.find_all(id="jornada19")

Transformamos a texto cada porción de el codigo html correspondiente a cada tabla

In [4]:
fecha1= str(fecha1) # transformamos a str la tabla
fecha2= str(fecha2)
fecha3= str(fecha3)
fecha4= str(fecha4)
fecha5= str(fecha5)
fecha6= str(fecha16)
fecha7= str(fecha7)
fecha8= str(fecha8)
fecha9= str(fecha9)
fecha10= str(fecha10)
fecha11= str(fecha11)
fecha12= str(fecha12)
fecha13= str(fecha13)
fecha14= str(fecha14)
fecha15= str(fecha15)
fecha16= str(fecha16)
fecha17= str(fecha17)
fecha18= str(fecha18)
fecha19= str(fecha19)

In [5]:
fecha1= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha1) #cambiamos a mayusculas
fecha2= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha2)
fecha3= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha3)
fecha4= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha4)
fecha5= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha5)
fecha6= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha6)
fecha7= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha7)
fecha8= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha8)
fecha9= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha9)
fecha10= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha10)
fecha11= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha11)
fecha12= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha12)
fecha13= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha13)
fecha14= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha14)
fecha15= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha15)
fecha16= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha16)
fecha17= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha17)
fecha18= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha18)
fecha19= re.sub(r'<.*?>', lambda g: g.group(0).upper(), fecha19)

Pasamos cada codigo a un dataframe correspondiente a cada fecha

In [6]:
dfecha1 = pd.read_html(fecha1)[0] 
dfecha2= pd.read_html(fecha2)[0]
dfecha3= pd.read_html(fecha3)[0]
dfecha4= pd.read_html(fecha4)[0]
dfecha5= pd.read_html(fecha5)[0]
dfecha6= pd.read_html(fecha6)[0]
dfecha7= pd.read_html(fecha7)[0]
dfecha8= pd.read_html(fecha8)[0]
dfecha9= pd.read_html(fecha9)[0]
dfecha10= pd.read_html(fecha10)[0]
dfecha11= pd.read_html(fecha11)[0]
dfecha12= pd.read_html(fecha12)[0]
dfecha13= pd.read_html(fecha13)[0]
dfecha14= pd.read_html(fecha14)[0]
dfecha15= pd.read_html(fecha15)[0]
dfecha16= pd.read_html(fecha16)[0]
dfecha17= pd.read_html(fecha17)[0]
dfecha18= pd.read_html(fecha18)[0]
dfecha19= pd.read_html(fecha19)[0]

Creamos una nueva columna de nombre fecha - donde le asignamos el numero de jornada a la que corresponde. Así podremos identificar más adelante a que jornada/fecha del campeonato corresponde cada partido

In [7]:
dfecha1['fecha']=1 #columna con el valor correspondiente a cada fecha.
dfecha2['fecha']=2
dfecha3['fecha']=3
dfecha4['fecha']=4
dfecha5['fecha']=5
dfecha6['fecha']=6
dfecha7['fecha']=7
dfecha8['fecha']=8
dfecha9['fecha']=9
dfecha10['fecha']=10
dfecha11['fecha']=11
dfecha12['fecha']=12
dfecha13['fecha']=13
dfecha14['fecha']=14
dfecha15['fecha']=15
dfecha16['fecha']=16
dfecha17['fecha']=17
dfecha18['fecha']=18
dfecha19['fecha']=19

Concatenamos todos los dataframe de fechas de manera de tener uno solo con todos los partidos jugados.

In [8]:
fixture2022 = pd.concat([dfecha1, dfecha2, dfecha3, dfecha4, dfecha5, dfecha6, dfecha7, dfecha8, dfecha9,
                             dfecha10, dfecha11, dfecha12, dfecha13, dfecha14, dfecha15,
                             dfecha16, dfecha17, dfecha18, dfecha19], axis=0) #concatenamos todos los dataframe fecha por fecha

In [9]:
fixture2022

Equipo local Resultado                       Equipo visitante  fecha
0     Barracas Central       1-1  Central Córdoba (Santiago del Estero)      1
1     Atlético Tucumán       1-1                         Colón Santa Fe      1
2          San Lorenzo       1-1                          Independiente      1
3             Banfield       1-2                      Newell's Old Boys      1
4            Patronato       1-1                        Vélez Sarsfield      1
..                 ...       ...                                    ...    ...
9                Lanús       0-1                           Boca Juniors     19
10  Defensa y Justicia       2-1                     Argentinos Juniors     19
11     Rosario Central       1-1                Estudiantes de La Plata     19
12    Atlético Tucumán       1-1                    Talleres de Córdoba     19
13   Gimnasia La Plata       2-0                     Arsenal de Sarandí     19

[266 rows x 4 columns]

252 rows y cuantro columnas

In [10]:
fixture2022.describe().T

count  mean      std  min  25%   50%   75%   max
fecha  266.0  10.0  5.48755  1.0  5.0  10.0  15.0  19.0

In [11]:
#fixture2022['Resultado']= fixture2022['Resultado'].astype(str)

Creamos dos columnas con goles de local y vistante, a partir de la columna resultado.

In [12]:
#creamos dos columnas una para gol visitante y otra para gol local seprando desde la columna Resultado tomando patron -
#tomando solo dos divisiones y con la opción expand true
fixture2022[['gol_local', 'gol_visitante']]= fixture2022.Resultado.str.split('-',2, expand=True)
fixture2022

Equipo local Resultado                       Equipo visitante  \
0     Barracas Central       1-1  Central Córdoba (Santiago del Estero)   
1     Atlético Tucumán       1-1                         Colón Santa Fe   
2          San Lorenzo       1-1                          Independiente   
3             Banfield       1-2                      Newell's Old Boys   
4            Patronato       1-1                        Vélez Sarsfield   
..                 ...       ...                                    ...   
9                Lanús       0-1                           Boca Juniors   
10  Defensa y Justicia       2-1                     Argentinos Juniors   
11     Rosario Central       1-1                Estudiantes de La Plata   
12    Atlético Tucumán       1-1                    Talleres de Córdoba   
13   Gimnasia La Plata       2-0                     Arsenal de Sarandí   

    fecha gol_local gol_visitante  
0       1         1             1  
1       1         1             1  
2       1         1             1  
3       1         1             2  
4       1         1             1  
..    ...       ...           ...  
9      19         0             1  
10     19         2             1  
11     19         1             1  
12     19         1             1  
13     19         2             0  

[266 rows x 6 columns]

descargamos el dataframe a un archivo csv

In [13]:
fixture2022.tail(20)

Equipo local Resultado                       Equipo visitante  \
8           San Lorenzo       1-1                       Atlético Tucumán   
9              Banfield       2-1                         Colón Santa Fe   
10   Arsenal de Sarandí       2-2                        Vélez Sarsfield   
11    Newell's Old Boys       2-0                      Gimnasia La Plata   
12         Boca Juniors       1-0                            River Plate   
13  Talleres de Córdoba       1-0                     Defensa y Justicia   
0             Sarmiento       1-2                          Independiente   
1       Vélez Sarsfield  Aplazado  Central Córdoba (Santiago del Estero)   
2            Godoy Cruz       1-1                                  Tigre   
3               Huracán       1-1                       Barracas Central   
4              Platense       1-0                         Unión Santa Fe   
5           Racing Club       1-0                              Patronato   
6              Aldosivi       1-2                      Newell's Old Boys   
7        Colón Santa Fe       0-0                            San Lorenzo   
8           River Plate       1-2                               Banfield   
9                 Lanús       0-1                           Boca Juniors   
10   Defensa y Justicia       2-1                     Argentinos Juniors   
11      Rosario Central       1-1                Estudiantes de La Plata   
12     Atlético Tucumán       1-1                    Talleres de Córdoba   
13    Gimnasia La Plata       2-0                     Arsenal de Sarandí   

    fecha gol_local gol_visitante  
8      18         1             1  
9      18         2             1  
10     18         2             2  
11     18         2             0  
12     18         1             0  
13     18         1             0  
0      19         1             2  
1      19  Aplazado          None  
2      19         1             1  
3      19         1             1  
4      19         1             0  
5      19         1             0  
6      19         1             2  
7      19         0             0  
8      19         1             2  
9      19         0             1  
10     19         2             1  
11     19         1             1  
12     19         1             1  
13     19         2             0

In [14]:
fixture2022.dtypes

Equipo local        object
Resultado           object
Equipo visitante    object
fecha                int64
gol_local           object
gol_visitante       object
dtype: object

nos quedaron salvo fecha del campeonato, todas las columnas como object - tanto gol de visitante y local deben de ser int64. Nos encontramos con el problema de que existen dos partidos aplazados y por ello no podemos cambiar el valor a númérico

In [15]:
aplazado= fixture2022[fixture2022.loc[:,'gol_local']== 'Aplazado'].index
fixture2022= fixture2022.drop(aplazado)

In [16]:
fixture2022

Equipo local Resultado                       Equipo visitante  \
0     Barracas Central       1-1  Central Córdoba (Santiago del Estero)   
2          San Lorenzo       1-1                          Independiente   
3             Banfield       1-2                      Newell's Old Boys   
4            Patronato       1-1                        Vélez Sarsfield   
5             Platense       2-1                             Godoy Cruz   
..                 ...       ...                                    ...   
9                Lanús       0-1                           Boca Juniors   
10  Defensa y Justicia       2-1                     Argentinos Juniors   
11     Rosario Central       1-1                Estudiantes de La Plata   
12    Atlético Tucumán       1-1                    Talleres de Córdoba   
13   Gimnasia La Plata       2-0                     Arsenal de Sarandí   

    fecha gol_local gol_visitante  
0       1         1             1  
2       1         1             1  
3       1         1             2  
4       1         1             1  
5       1         2             1  
..    ...       ...           ...  
9      19         0             1  
10     19         2             1  
11     19         1             1  
12     19         1             1  
13     19         2             0  

[247 rows x 6 columns]

In [17]:
fixture2022['gol_local'] = fixture2022['gol_local'].astype(np.int64) #le asignamos tipo valor float
fixture2022['gol_visitante'] = fixture2022['gol_visitante'].astype(np.int64)
fixture2022

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Equipo local Resultado                       Equipo visitante  \
0     Barracas Central       1-1  Central Córdoba (Santiago del Estero)   
2          San Lorenzo       1-1                          Independiente   
3             Banfield       1-2                      Newell's Old Boys   
4            Patronato       1-1                        Vélez Sarsfield   
5             Platense       2-1                             Godoy Cruz   
..                 ...       ...                                    ...   
9                Lanús       0-1                           Boca Juniors   
10  Defensa y Justicia       2-1                     Argentinos Juniors   
11     Rosario Central       1-1                Estudiantes de La Plata   
12    Atlético Tucumán       1-1                    Talleres de Córdoba   
13   Gimnasia La Plata       2-0                     Arsenal de Sarandí   

    fecha  gol_local  gol_visitante  
0       1          1              1  
2       1          1              1  
3       1          1              2  
4       1          1              1  
5       1          2              1  
..    ...        ...            ...  
9      19          0              1  
10     19          2              1  
11     19          1              1  
12     19          1              1  
13     19          2              0  

[247 rows x 6 columns]

In [18]:
fixture2022.dtypes

Equipo local        object
Resultado           object
Equipo visitante    object
fecha                int64
gol_local            int64
gol_visitante        int64
dtype: object

Debemos ahora crear una columna que nos indique el resultado. esto es, si gana el local, el visitante o si empataron

In [19]:
#definimos las condiciones que deben darse para uno u otro resultado

condiciones= [fixture2022['gol_local']>fixture2022['gol_visitante'], fixture2022['gol_local']<fixture2022['gol_visitante']]

# definimos las opciones
opciones= ['local', 'visitante']

#creamos la columna del dataframe que muestre el resultado de la comparación
fixture2022['resultado'] =np.select(condiciones, opciones, 'empate')

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if __name__ == '__main__':


In [20]:
fixture2022= fixture2022.rename(columns={'Resultado':'marcador'}) #cambiamos la columna del marcador para no confundir el resultado
fixture2022

Equipo local marcador                       Equipo visitante  fecha  \
0     Barracas Central      1-1  Central Córdoba (Santiago del Estero)      1   
2          San Lorenzo      1-1                          Independiente      1   
3             Banfield      1-2                      Newell's Old Boys      1   
4            Patronato      1-1                        Vélez Sarsfield      1   
5             Platense      2-1                             Godoy Cruz      1   
..                 ...      ...                                    ...    ...   
9                Lanús      0-1                           Boca Juniors     19   
10  Defensa y Justicia      2-1                     Argentinos Juniors     19   
11     Rosario Central      1-1                Estudiantes de La Plata     19   
12    Atlético Tucumán      1-1                    Talleres de Córdoba     19   
13   Gimnasia La Plata      2-0                     Arsenal de Sarandí     19   

    gol_local  gol_visitante  resultado  
0           1              1     empate  
2           1              1     empate  
3           1              2  visitante  
4           1              1     empate  
5           2              1      local  
..        ...            ...        ...  
9           0              1  visitante  
10          2              1      local  
11          1              1     empate  
12          1              1     empate  
13          2              0      local  

[247 rows x 7 columns]

In [21]:
fixture2022.to_csv('fixtureliga_arg_2022.csv', encoding='utf-8') # pasamos el dataframe a un archivo csv

In [22]:
fixture2022.describe().T

count       mean       std  min  25%   50%   75%   max
fecha          247.0  10.000000  5.488347  1.0  5.0  10.0  15.0  19.0
gol_local      247.0   1.210526  1.022218  0.0  0.0   1.0   2.0   5.0
gol_visitante  247.0   0.874494  0.908652  0.0  0.0   1.0   1.0   5.0

Como podemos ver - los equipos locales tienen un promedio de 1.21 goles por partido y los visitantes de 0.87 - 

In [38]:
fixture2022.describe(include='O').T

count unique                      top freq
Equipo local       247     28  Estudiantes de La Plata   11
marcador           247     21                      1-1   43
Equipo visitante   247     28          Vélez Sarsfield   11
resultado          247      3                    local  105

el resultado que más se dió es el 1-1 (43 veces) 
de 247 partidos disputados - 105 gano el local. Siendo el resultado con mayor frecuencia.

Generamos una tabla por equipo, con los resultados cuando jugo de LOCAL

In [36]:
fixture2022.pivot_table(index='Equipo local', columns='resultado', aggfunc={"resultado": len})
# aggfunc={"gol_local": sum, "gol_visitante": sum}

resultado                
resultado                                empate local visitante
Equipo local                                                   
Aldosivi                                    1.0   1.0       6.0
Argentinos Juniors                          1.0   6.0       3.0
Arsenal de Sarandí                          5.0   1.0       1.0
Atlético Tucumán                            3.0   5.0       NaN
Banfield                                    3.0   2.0       5.0
Barracas Central                            2.0   5.0       3.0
Boca Juniors                                1.0   8.0       1.0
Central Córdoba (Santiago del Estero)       2.0   4.0       2.0
Colón Santa Fe                              4.0   3.0       3.0
Defensa y Justicia                          6.0   3.0       1.0
Estudiantes de La Plata                     3.0   7.0       1.0
Gimnasia La Plata                           2.0   5.0       NaN
Godoy Cruz                                  3.0   5.0       NaN
Huracán                                     3.0   3.0       1.0
Independiente                               4.0   3.0       2.0
Lanús                                       3.0   2.0       3.0
Newell's Old Boys                           3.0   3.0       3.0
Patronato                                   3.0   5.0       NaN
Platense                                    6.0   3.0       NaN
Racing Club                                 3.0   6.0       1.0
River Plate                                 1.0   5.0       3.0
Rosario Central                             2.0   5.0       2.0
San Lorenzo                                 9.0   2.0       NaN
Sarmiento                                   2.0   2.0       2.0
Talleres de Córdoba                         3.0   4.0       3.0
Tigre                                       5.0   3.0       1.0
Unión Santa Fe                              3.0   3.0       3.0
Vélez Sarsfield                             4.0   1.0       2.0

Generamos una tabla por equipo, con los resultados cuando jugo de VISITANTE

In [37]:
fixture2022.pivot_table(index='Equipo visitante', columns='resultado', aggfunc={"resultado": len})

resultado                
resultado                                empate local visitante
Equipo visitante                                               
Aldosivi                                    4.0   5.0       1.0
Argentinos Juniors                          2.0   5.0       1.0
Arsenal de Sarandí                          5.0   4.0       NaN
Atlético Tucumán                            4.0   3.0       2.0
Banfield                                    3.0   2.0       3.0
Barracas Central                            4.0   4.0       1.0
Boca Juniors                                1.0   4.0       4.0
Central Córdoba (Santiago del Estero)       2.0   3.0       2.0
Colón Santa Fe                              2.0   5.0       1.0
Defensa y Justicia                          1.0   4.0       3.0
Estudiantes de La Plata                     3.0   4.0       1.0
Gimnasia La Plata                           5.0   3.0       2.0
Godoy Cruz                                  2.0   5.0       3.0
Huracán                                     5.0   1.0       3.0
Independiente                               3.0   5.0       2.0
Lanús                                       3.0   7.0       NaN
Newell's Old Boys                           1.0   3.0       5.0
Patronato                                   1.0   4.0       1.0
Platense                                    2.0   4.0       3.0
Racing Club                                 4.0   3.0       1.0
River Plate                                 5.0   2.0       3.0
Rosario Central                             4.0   3.0       1.0
San Lorenzo                                 4.0   1.0       3.0
Sarmiento                                   2.0   6.0       2.0
Talleres de Córdoba                         4.0   4.0       NaN
Tigre                                       4.0   4.0       1.0
Unión Santa Fe                              3.0   3.0       3.0
Vélez Sarsfield                             7.0   4.0       NaN